auth/authlink/store.go

```diff
@@ -0,0 +1,240 @@
+package authlink
+
+import (
+	"context"
+	"crypto/rand"
+	"database/sql"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/jackc/pgtype"
+	uuid "github.com/satori/go.uuid"
+	"github.com/target/goalert/keyring"
+	"github.com/target/goalert/permission"
+	"github.com/target/goalert/util"
+)
+
+const (
+	MaxClaimQueryFrequency = 3 * time.Second
+	ExpirationDuration     = 10 * time.Minute
+)
+
+type Store struct {
+	db *sql.DB
+
+	unclaimed *sql.Stmt
+
+	create *sql.Stmt
+	claim  *sql.Stmt
+	verify *sql.Stmt
+	auth   *sql.Stmt
+
+	reset  *sql.Stmt
+	status *sql.Stmt
+
+	mx           sync.Mutex
+	closeCh      chan struct{}
+	waitRefresh  chan struct{}
+	startRefresh chan struct{}
+	wl           *Whitelist
+
+	keys keyring.Keyring
+}
+
+type Status struct {
+	ID string
+
+	ClaimCode  string
+	VerifyCode string
+
+	CreatedAt  time.Time
+	ExpiresAt  time.Time
+	ClaimedAt  time.Time
+	VerifiedAt time.Time
+	AuthedAt   time.Time
+}
+
+const codeAlphabet = "ABCDEFGHJKLMNPQRSTWXYZ23456789"
+
+func genCode(n int) (string, error) {
+	val := make([]byte, n)
+	for i := range val {
+		num, err := rand.Int(rand.Reader, big.NewInt(int64(len(codeAlphabet))))
+		if err != nil {
+			return "", err
+		}
+		val[i] = codeAlphabet[num.Int64()]
+	}
+
+	return string(val), nil
+}
+
+func NewStore(ctx context.Context, db *sql.DB, keys keyring.Keyring) (*Store, error) {
+	p := util.Prepare{DB: db, Ctx: ctx}
+	s := &Store{
+		wl:           NewWhitelist(),
+		keys:         keys,
+		closeCh:      make(chan struct{}),
+		waitRefresh:  make(chan struct{}),
+		startRefresh: make(chan struct{}),
+
+		unclaimed: p.P(`
+			select claim_code
+			from auth_link_codes
+			where
+				now() < expires_at and
+				claimed_at isnull and
+				verified_at isnull and
+				authed_at isnull
+		`),
+
+		create: p.P(`
+			insert into auth_link_codes (id, user_id, auth_user_session_id, claim_code, verify_code, expires_at)
+			values ($1, $2, $3, $4, $5, now() + $6::interval)
+			returning created_at
+		`),
+		claim: p.P(`
+			update auth_link_codes set claimed_at = now()
+			where
+				claim_code = $1 and
+				now() < expires_at and
+				claimed_at isnull and
+				verified_at isnull and
+				authed_at isnull
+			returning id, claimed_at, verify_code, expires_at
+		`),
+		verify: p.P(`
+			update auth_link_codes
+			set verified_at = now()
+			where
+				id = $1 and
+				user_id = $2 and
+				auth_user_session_id = $3
+				and verify_code = $4 and
+				now() < expires_at and
+				claimed_at notnull and
+				verified_at isnull and
+				authed_at isnull
+		`),
+		auth: p.P(`
+			update auth_link_codes
+			set authed_at = now()
+			where
+				id = $1 and
+				now() < expires_at and
+				claimed_at notnull and
+				verified_at notnull and
+				authed_at isnull
+			returning user_id
+		`),
+
+		reset: p.P(`delete from auth_link_codes where user_id = $1`),
```

Comment:
Is this intended to reset all sessions for a user, or should this be `where id = $1` to reset a single session? If the former, could we perhaps rename to `resetAll` for clarity?

Reply:
A user can only link a single device at a time, and linking is locked to a specific session so there will only be 0 or 1, this reset's the user's link attempt, but is using the user_id instead of session in case it was started in another tab/browser